granulate_utils/linux/devkmsg.py

```diff
@@ -0,0 +1,58 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+import errno
+import os
+import time
+from typing import List, Tuple
+
+from granulate_utils.linux.kernel_messages import KernelMessagesProvider
+
+# See linux/printk.h
+CONSOLE_EXT_LOG_MAX = 8192
+
+
+class DevKmsgProvider(KernelMessagesProvider):
+    """
+    Provides a stream of kernel messages via /dev/kmsg. Requires Linux 3.5+.
+
+    The /dev/kmsg interface is described at Documentation/ABI/testing/dev-kmsg in the kernel source tree
+    and can be viewed at https://github.com/torvalds/linux/blob/master/Documentation/ABI/testing/dev-kmsg.
+    """
+
+    def __init__(self):
+        self.dev_kmsg_fd = os.open("/dev/kmsg", os.O_RDONLY)
```

Comment:
Open it with `open` and store the result, then use `.fileno()`. This way the result is cleaned up automatically once this object dies.